### DISCUSSION TITLE
Properly starting async connection with SQLAlchemy Core

### DISCUSSION THREAD

**Discussion by jocho-here** [fastapi/fastapi#4302] [Questions] [answered]
I'm trying to build an async backend API service using FastAPI, Postgres, and SQLAlchemy Core.

When I was testing my API service without the frontend, it ran perfectly, but as soon as the frontend started to hit the backend many times concurrently, I started seeing the following error:
```
ERROR:    Exception in ASGI application
...
sqlalchemy.exc.ResourceClosedError: This Connection is closed
```

Here's my setup:

1. service/utils/db.py
```
from fastapi import Request
from sqlalchemy.ext.asyncio import create_async_engine
from service.utils.config import DB_DSN
...
def get_async_engine(echo: bool = True, future: bool = Ture):
    global ENGINE

    if not ENGINE:
        ENGINE = create_async_engine(
            str(DB_DSN), echo=echo, future=future
        )

    return ENGINE

async def get_async_conn():
    engine = get_async_engine()

    async with engine.begine() as conn:
        Request.db_conn = conn
        yield
```

2. service/api/main.py
```
from fastapi import Depends, FastAPI

from service.utils.db import get_async_conn

def get_app():
    app = FastAPI(
        title="app", dependencies=[Depends(get_async_conn)]
    )
    ...

    return app
```

3. Example endpoint
```
from fastapi import APIRouter

router = APIRouter(...)

@router.get("")
async def get_items():
    query = await Request.db_conn.execute(...)  # some SQL query
    items = query.fetchall()
```

Any suggestion on how to **properly** manage async engine & connection with SQLAlchemy Core?

### EDIT
I am not trying to use SQLAlchemy's ORM functionality, but raw SQL on SQLAlchemy engine.

**Comment by bluebrown** [comment, marked as answer]
I did it like this. 3 files. I haven't really done any stress testing. So I hope it will not happen the same to me.

db.init.py

```python
from sqlalchemy.ext.asyncio import AsyncEngine, AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker

from backend.core.config import POSTGRES_URL

engine: AsyncEngine
session: AsyncSession


def init_engine_and_session() -> AsyncEngine:
    global engine, session
    engine = create_async_engine(str(POSTGRES_URL))
    session = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)
    return engine
```

main.py

```python
from fastapi import FastAPI

def create_app() -> FastAPI:
    import backend.db.init as postgres

    root_app = FastAPI()
    engine = postgres.init_engine_and_session()
    root_app.add_event_handler("shutdown", engine.dispose)
```

deps.py

```python
from typing import Any, AsyncGenerator
from sqlalchemy.ext.asyncio import AsyncSession

import backend.db.init as postgres


async def session() -> AsyncGenerator[AsyncSession, Any]:
    async with postgres.session() as session:
        async with session.begin():
            yield session
```

Then I use uvicorn with factory flag to start the app

```bash
uvicorn backend.main:create_app --factory
```

**Reply by jocho-here** [reply]
Hmmm, I'm looking for an answer that's using SQLAlchemy Core's async engine, instead of their session. With your suggestion, I got a connection out of the session (after `session.begin()`) and passed it to my `Request.db_conn` and used it. It worked just like how it would work as how I'm using it right now (bare `engine`).

Question, have you run your backend with a running frontend application that makes multiple concurrent requests? I'm trying to see if running multiple processes of API servers using `gunicorn` with `uvicorn` would resolve this issue.

**Reply by grillazz** [reply]
@mgjo5899 take a look here https://github.com/grillazz/fastapi-sqlalchemy-asyncpg

**Reply by bluebrown** [reply]
@mgjo5899 you dont have to use a session to get a connection.

```python
async def connection() -> AsyncGenerator[AsyncConnection, None]:
    async with postgres.engine.connect() as conn:
        yield conn
```

Regarding client connection. I don't have a frontend yet, but after I read your question, I have used [autocannon](https://github.com/mcollina/autocannon) to do tests.

It seems like it can handle the traffic OK from concurrent client. But as soon as you start [pipelining](https://en.wikipedia.org/wiki/HTTP_pipelining), errors fly all around. I assume this is more on the fastapi side than sqlalchemy.

When using pipelining, it immediately throws this error repeatedly. I assume uvicorn or fastapi simply can't handle pipelines.

```bash
RuntimeError: unable to perform operation on <TCPTransport closed=True reading=False 0x7f3f1b8aeff0>; the handler is closed
ERROR: Exception in ASGI application
```

Without pipelining and 100 concurrent connections, it works, but not as stable as I would like to. I have observed this error a few times, which could potentially get tweaked by configuring the engine differently. I have looked into that yet.

```bash
sqlalchemy.exc.TimeoutError: QueuePool limit of size 5 overflow 10 reached, connection timed out, timeout 30.00
```

Apart from that, I have to say that it's not as performant as advertised. In the docs/readme, it's told that it would be on par with nodejs or even golang. I have written a server in go that connects to a postgresdb with the same schema as this fastapi backend, and implements the same handler that was tested on fastapi. 
FastAPI is able to handle, about 3k requests in 10 seconds. Golang was able to handle 16k requests in 10 seconds and even up to 26k with pipelining (each connection pipelines 10 requests).
